chain/network/src/tests/cache_edges.rs

```diff
@@ -1,311 +1,221 @@
-            let peer0 = self.get_peer(*peer0).clone();
-            let peer1 = self.get_peer(*peer1).clone();
-            let (peer0, peer1) = Edge::make_key(peer0, peer1);
-
-            let res = self.routing_table.edges_info.get(&(peer0, peer1));
-            assert!(res.is_some());
-            let edge = res.unwrap();
-            assert_eq!(edge.edge_type(), *edge_type);
-        }
-        let active_edges = on_memory
-            .iter()
-            .filter_map(|x| if x.2 == EdgeState::Active { Some(1) } else { None })
-            .count();
-
-        assert_eq!(active_edges, self.routing_table.raw_graph.total_active_edges() as usize);
-        assert_eq!(
-            active_edges,
-            self.routing_table.raw_graph.compute_total_active_edges() as usize
-        );
-        assert_eq!(on_memory.len(), self.routing_table.edges_info.len());
-
-        // Check for peers on disk
-        let mut total_peers = 0;
-        for (peer, nonce) in self.store.iter(DBCol::PeerComponent) {
-            total_peers += 1;
-
-            let peer = PeerId::try_from_slice(peer.as_ref()).unwrap();
-            let nonce = u64::try_from_slice(nonce.as_ref()).unwrap();
-            let peer_ix = self.rev_peers.get(&peer).unwrap();
-            let res = on_disk_peers.get(peer_ix).unwrap();
-            assert_eq!(*res, nonce);
-        }
-        assert_eq!(total_peers, on_disk_peers.len());
-
-        // Check for edges on disk
-        let mut total_nonces = 0;
-        for (nonce, edges) in self.store.iter(DBCol::ComponentEdges) {
-            total_nonces += 1;
-
-            let nonce = u64::try_from_slice(nonce.as_ref()).unwrap();
-
-            let edges = Vec::<Edge>::try_from_slice(edges.as_ref()).unwrap();
-            let current_edges = on_disk_edges.get(&nonce).unwrap();
-
-            assert_eq!(edges.len(), current_edges.len());
+    fn new() -> Self {
+        let mut rng = util::make_rng(87927345);
+        let clock = time::FakeClock::default();
+        let me = data::make_peer_id(&mut rng);
+        let store = create_test_store();
 
-            for edge in edges.iter() {
-                let edge_description = self.get_edge_description(edge);
-                assert!(current_edges.contains(&edge_description));
-            }
-        }
-        assert_eq!(total_nonces, on_disk_edges.len());
+        let graph = Arc::new(RwLock::new(routing::GraphWithCache::new(me.clone())));
+        Self { rng, clock, graph, store, _system: actix::System::new() }
     }
 
-    fn add_edge(&mut self, peer0: usize, peer1: usize, nonce: u64) {
-        let peer0 = self.get_peer(peer0).clone();
-        let peer1 = self.get_peer(peer1).clone();
-        let edge = Edge::new(peer0, peer1, nonce, Signature::default(), Signature::default());
-        self.routing_table.add_verified_edges_to_routing_table(vec![edge]);
+    fn component_store(&self) -> routing::ComponentStore {
```

Comment:
why not? Isn't Store stateless?

Reply:
near_store::Store is cloneable and used by multiple actors. What would you like to gain here?

Reply:
> What would you like to gain here?

You'd gain a local invariant -- that the store part which relates to the network has a unique owner, and doesn't have to care about DB-level races. This is a weak invariant in a sense that that if you go out of your way and clone the raw store (or just run a second process with the same DB), you can circumvent the invariant.

It still seems a useful invariant to me:

* while not strictly water-tight, it would be hard to violate it by accident
* "there's no concurrency shenanigans happening in this corner of the database" seems like a particularly useful idea to communicate to the reader.

Reply:
Ok, I've merged the "store views" of all modules in near-network (there is no ComponentStore any more, as discussed on one of your other comments), so now multiple objects contain a reference to the same storage. The invariant that I wanted to introduce here was that every method of the Store struct corresponds to a single commit (aka DB transaction) and should be treated as an atomic operation (modulo the fact that currently you cannot perform transactional reads). 

What you propose is an equivalent of a compile-time mutex on all DB transactions when using a single handler. The alternative would be a runtime mutex either on the whole DB or a more granular one (basically to support placing reads within a transaction). IMO we should rather go in this direction. How does it look like in the other regions of our codebase? (like, the stuff that you were refactoring recently) I recall that there was an idea for implementing per row locking for one of the use cases.

Reply:
>What you propose is an equivalent of a compile-time mutex on all DB transactions when using a single handler

Yeah, exactly, that's much better way to describe it.


>How does it look like in the other regions of our codebase?

In chain at least, this is exactly the opposite of "every method of the Store struct corresponds to a single commit". The chain generally wants to batch up changes from many separate methods into a single transaction and apply that atomically. So various components don't commit themselves but rather return updates, than those updates are merged into a single update, and than that is commited (or not) atomically.

I think we don't necessary need to have one unified pattern for the whole codebase -- some components might need transactions spanning complex operations, some components might want to have a fine-grained "method=immediate commit" semantics. 

The only general guideline I'd love to aspire to is to minimize the spatial and temporal extent where access to the database happens. From readability and unserstandability perspective, it's very useful if you can point at a single narrow waist where all DB modifications happen. From performance perspective, sometimes you have to do db accesses, and sometimes even writes, in parallel (the narrow waist could be a bottleneck), but doing that increases complexity a lot. 

In other words, the underlying DB API allows for concurrency, but I'd rather not take advantage of that unless we absolutely must.

Reply:
ok, I've made crate::store::Store non-cloneable and postponed wrapping near_store::Store long enough to clone and distribute the handlers to all components (it is a bit more verbose now). Another thing I could do would be to pass around store handler the same way as the clock (then there would be just 1 store handler per actor) but again it would get even more verbose. WDYT?